_drafts/assert.md

```diff
@@ -0,0 +1,605 @@
+configurable on mature software. One example of configurable asserting in
+library code is the `xTaskCreateStatic` function in FreeRTOS[^4].
+
+In this function, `configASSERT` is used to check for non-null pointers, and if
+`configASSERT` is configured to a no-op, the code will still check those values
+later in the function and return an error if they are invalid.
+
+### When undefined behavior is immenent
+
+```c
+typedef enum {
+  EventType_A,
+  EventType_B,
+} EventType;
+
+void handle_event(Event *evt) {
+  ...
+  switch (evt->type) {
+    case EventType_A:
+      // something
+    case EventType_B:
+      // something
+    default:
+      // shouldn't ever get here.
+      // something is corrupted
+      ASSERT(0);
+  }
+}
+```
+
+If we want to make this even more understandable for future developers who come
+across this, something like the following can also be done.
+
+```c
+#define WTF ASSERT(0)
+
+
+void handle_event(Event *evt) {
+  ...
+  switch (evt->type) {
+    ...
+    default:
+      WTF;
+  }
+}
+```
+
+This makes it pretty obvious that anything that falls through to the `default:`
+case is undefined behavior.
+
+### RTOS API functions fail
+
+Another reasonable place to assert is when you expect certain system function to
+succeed, and if they don't, the system is in a bad state. Below is a snippet
+where we assert if a mutex wasn't locked with a few seconds, which likely means
+there is a deadlock somewhere else.
+
+```c
+int rv = xSemaphoreTake(s_mutex, sec_to_ticks(5));
+ASSERT(rv == pdTRUE);
+```
+
+## When not to assert
+
+There are places where it is advised to assert. On the contrary, there are
+places where it's critical to **not** assert.
+
+- Avoid asserting on boot up sequences. This is the most common way to introduce
+  reboot loops.
+- Don't assert on operations that depend on the hardware behaving appropriately.
+  If a sensor says it will return a value between 0-100, it's probably best not
+  assert when it's above 100, because you can never trust today's cheap
+  hardware.
+- Don't assert on the contents of data read from persistent storage, unless it's
+  guaranteed to be valid. The data read from flash or a filesystem could be
+  corrupted.
+- If it is very likely the system will recover in a few moments, it might be
+  best to not assert.
+
+## Custom Assert Implementation Examples
+
+It's now time to start asserting ourselves. The examples provided below progress
+from a naive solution to the best one I've seen and used to date, with each
+becoming more expressive while also using less (precious) code space.
+
+> NOTE: All of these implementations will behave similarly if a debugger is
+> attached. The goal of the exercise is to improve upon asserts that are hit
+> while a debugger is **not** attached and all we have is a logging system.
+
+The code starts out in `prv_main`, and will call one of two functions,
+`prv_one_true`, or `prv_one_false`, depending on a randomly generated boolean
```

Comment:
nit: I don't quite understand what the names mean, but not a big deal. I think prv_assert_path_one, prv_assert_path_two, etc may have made more sense to me

Reply:
That makes more sense. I'll change them.